src/v/kafka/server/handlers/produce.cc

```diff
@@ -234,6 +234,77 @@ ss::future<produce_response::partition> finalize_request_with_error_code(
         .partition_index = ntp.tp.partition, .error_code = ec});
 }
 
+/**
+ * @brief Validate the timestamps of the batch, they have to be within a window
+ * to the broker's time. returns the new timestamp to set as max_timestamp to
+ * the batch, if present
+ */
+static auto validate_batch_timestamps(
+  model::ntp const& ntp,
+  model::record_batch const& batch,
+  model::timestamp_type timestamp_type,
+  net::server_probe& probe) -> std::optional<model::timestamp> {
+    // we compute in std::chrono::timepoints, we print in model::timestamps
+    auto broker_time = model::timestamp::now();
+    auto broker_timepoint = model::duration_since_epoch(broker_time);
+
+    // alert if first_timestamp is too far in the past
+    // default value for threshold basically disables this check, so it's
+    // wrapped in a if check
```

Comment:
question: Do we support std::optional for config values?

Reply:
we do, I went for the symmetry of the __after__ property not being optional, but i can change it

Reply:
Which ever is cleaner in your mind, I don't feel strongly but was more asking for my knowledge :)